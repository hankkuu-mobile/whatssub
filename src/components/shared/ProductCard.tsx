import { IC_EDIT, IC_NOTI_OFF, IC_NOTI_ON } from '../../utils/Icons';
import React, { FC, useState } from 'react';
import { SwitchToggle } from 'dooboo-native-widgets';
import {
  TouchableOpacity,
} from 'react-native';

import { getString } from '../../../STRINGS';
import moment from 'moment';
import styled from 'styled-components/native';

const Container = styled.View<{ isSummary: boolean }>`
  display: flex;
  position: relative;
  width: 100%;
  height: ${({ isSummary }) => isSummary ? '82' : '190'};
  background-color: white;
  padding: 15px 22px;
  border-radius: 15;
  
`;

const HeaderView = styled.View<{ isSubscription: boolean }>`
  position: absolute;
  width: 100%;
  height: ${({ isSubscription }) => isSubscription ? '40' : '30'};
  top: 15;
  left: 22;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const ContentView = styled.View`
  display: flex;
`;

const SummaryContentView = styled.View`
  display: flex;
  width: 100%;
  height: 100%;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const SummaryContentViewWrapper = styled.View`
  position: absolute;
  left: 20;
  top: 30;
  width: 100%;
`;

const TitleText = styled.Text`
  font-family: 'spoqa-han-sans-bold';
  margin-top: 17;
  text-align: center;
  font-size: 16;
  font-weight: bold;
  color: ${({ theme }) => theme.fontColor};
`;

const SummaryTitleTextWrapper = styled.View`
  display: flex;
`;

const SummaryTitleText = styled.Text`
  font-family: 'spoqa-han-sans-bold';
  text-align: center;
  font-size: 16;
  font-weight: bold;
  color: ${({ theme }) => theme.fontColor};
`;

const PriceText = styled.Text`
  font-family: 'spoqa-han-sans-regular';
  font-size: 14;
  text-align: center;
  color: rgb(153, 162, 170);
  margin-top: 10;
`;

const DateText = styled.Text<{ variant: Variant, isSummary: boolean }>`
  font-family: ${({ isSummary }) =>
    isSummary ? 'spoqa-han-sans-bold' : 'spoqa-han-sans-regular'};
  font-size: ${({ isSummary }) =>
    isSummary ? '16' : '40'};
  text-align: center;
  color: ${({ variant }) =>
    variant === 'subscription'
      ? 'rgb(255, 77, 108)'
      : 'rgb(62, 126, 255)'};
`;

const ProductImage = styled.Image<{ isSubscription: boolean }>`
  width: ${({ isSubscription }) => isSubscription ? '75' : '31'};
  height: ${({ isSubscription }) => isSubscription ? '20' : '31'};
  margin-top: ${({ isSubscription }) => isSubscription ? '10' : '0'};
  margin-bottom: ${({ isSubscription }) => isSubscription ? '10' : '0'};
  margin-right: auto;
  margin-left: auto;
`;

const IconImage = styled.Image`
  width: 24;
  height: 24;
`;

const IconWrapper = styled.View`
  display: flex;
  flex-direction: row;
`;

const ContentText = styled.Text`
  font-family: 'spoqa-han-sans-bold';
  text-align: center;
  font-size: 13;  
`;

export enum Variant {
  Subscription = 'subscription',
  Payment = 'payment',
};

export interface ProductCardProps {
  onClickNotification: () => void;
  onClickSubscription?: () => void;
  onClickEdit: () => void;
  isNotificationEnable: boolean;
  isSubscribing?: boolean;
  name: string;
  image: string;
  price: number;
  currentMonthPaymentDate: Date;
  variant?: Variant;
  defaultIsSummary?: boolean;
}

const ProductCard: FC<ProductCardProps> = ({
  name, image, price, currentMonthPaymentDate, variant = Variant.Subscription,
  isNotificationEnable, isSubscribing = false, onClickSubscription, onClickNotification,
  onClickEdit, defaultIsSummary = false,
}) => {
  const dDay = moment(currentMonthPaymentDate)
    .diff(moment().startOf('day'), 'day');
  const isSubscription = variant === Variant.Subscription;
  const [isSummary, setIsSummary] = useState(defaultIsSummary);

  const onPressCard = () => {
    setIsSummary((prev) => !prev);
  };

  const onClickSwitchToggle = () => {
    onClickSubscription();
    return null;
  };

  const onPressNotiIcon = () => {
    onClickNotification();
  };

  const renderNoti = isNotificationEnable
    ? <TouchableOpacity onPress={onPressNotiIcon} testID='notiOffIcon'>
      <IconImage source={IC_NOTI_OFF}/>
    </TouchableOpacity>
    : <TouchableOpacity onPress={onPressNotiIcon} testID="notiOnIcon">
      <IconImage source={IC_NOTI_ON}/>
    </TouchableOpacity>;

  const renderSwitch = <SwitchToggle
    backTextRight={isSubscribing
      ? '' : getString('PRODUCTION_CARD_SUBSCRIPTION_SWITCH_OFF')}
    backTextLeft={isSubscribing
      ? getString('PRODUCTION_CARD_SUBSCRIPTION_SWITCH_ON') : ''}
    textRightStyle={{ color: 'white', position: 'absolute', left: 5, top: -7 }}
    textLeftStyle={{ color: 'white', position: 'absolute', left: 5, top: -7 }}
    switchOn={isSubscribing}
    backgroundColorOn={'rgb(2, 48, 89)'}
    backgroundColorOff={'rgb(2, 48, 89)'}
    circleColorOn={'white'}
    circleColorOff={'white'}
    onPress={onClickSwitchToggle}/>;

  return (
    <TouchableOpacity onPress={onPressCard} testID='productCard'>
      <Container isSummary={isSummary}>
        {isSummary
          ? <>
            <SummaryTitleTextWrapper testID='summaryTitleText'>
              <SummaryTitleText>{name}</SummaryTitleText>
              <PriceText>{price && `-${price.toLocaleString()}원`}</PriceText>
            </SummaryTitleTextWrapper>
            <SummaryContentViewWrapper>
              <SummaryContentView>
                <DateText variant={variant} isSummary={isSummary}>
                  {`D-${dDay}`}
                </DateText>
                {isSubscription ? renderSwitch : renderNoti}
              </SummaryContentView>
            </SummaryContentViewWrapper>
          </>
          : <>
            <ProductImage
              testID='productImg'
              source={{ uri: image }}
              isSubscription={isSubscription}
            />
            <HeaderView isSubscription={isSubscription}>
              <IconWrapper>
                <TouchableOpacity onPress={onClickEdit}>
                  <IconImage source={IC_EDIT}/>
                </TouchableOpacity>
                {isSubscription && renderNoti}
              </IconWrapper>
              {isSubscription ? renderSwitch : renderNoti}
            </HeaderView>
            <ContentView>
              <TitleText>{name}</TitleText>
              <DateText variant={variant} isSummary={false}>
                {`D-${dDay}`}
              </DateText>
              <ContentText>
                {`${moment(currentMonthPaymentDate).format('M월 D일')}에
                  ${price && price.toLocaleString()}원이 결제될 예정입니다.`}
              </ContentText>
            </ContentView>
          </>
        }
      </Container>
    </TouchableOpacity>
  );
};

export default ProductCard;
