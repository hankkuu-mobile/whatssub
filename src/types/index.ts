import { DefaultTheme } from 'styled-components';
/**
 * Fix error typing in `split` method in `apollo-link`
 * Related issue https://github.com/apollographql/apollo-client/issues/3090
 */
export interface Definition {
  kind: string;
  operation?: string;
};

export enum Gender { MALE, FEMALE };

export interface SocialInput {
  social: string;
  photo?: string;
  email?: string;
  password?: string;
  name?: string;
  nickname?: string;
  birthday?: Date;
  gender?: Gender;
  phone?: string;
}

export interface User {
  id: string;
  email: string;
  name: String;
  nickname: String;
  photo: String;
  birthday: Date;
  gender: Gender;
  social: String;
  verified: Boolean;
  notifications: [Notification]
  // customServices: [CustomService]
  // userProducts: [UserProduct]
  // reviews: [Review]
  created: Date
  updated: Date
  deleted: Date
}

export interface AuthPayload {
  token: string
  user: User
}

export enum ThemeType {
  LIGHT = 'LIGHT',
  DARK = 'DARK',
};

export interface ScreenProps {
  theme: DefaultTheme;
  changeTheme: () => void;
}
