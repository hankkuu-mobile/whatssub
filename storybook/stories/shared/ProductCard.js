import ProductCard, {
  Variant,
} from '../../../src/components/shared/ProductCard';
import React, { useState } from 'react';

import { storiesOf } from '@storybook/react-native';
import styled from 'styled-components/native';

const Container = styled.SafeAreaView`
  flex: 1;
`;
const ContainerDeco = (storyFn) => <Container>{storyFn()}</Container>;

const Story = (props) => {
  const [isNotificationEnable1, setNotificationEnable1] = useState(
    props.notification ? props.notification : false,
  );
  const [isSubscribing1, setSubscribing1] = useState(
    props.subscribing ? props.subscribing : false,
  );

  return (
    <ProductCard
      onClickNotification={() => setNotificationEnable1((prev) => !prev)}
      onClickSubscription={() => setSubscribing1((prev) => !prev)}
      onClickEdit={() => {}}
      isNotificationEnable={isNotificationEnable1}
      isSubscribing={isSubscribing1}
      name='netflix'
      image='https://dummyimage.com/75x20/000/fff'
      price={65000}
      currentMonthPaymentDate={new Date('2019-7-28')}
      variant={Variant.Subscription}
      defaultIsSummary={props.summary ? props.summary : false}
    />
  );
};

storiesOf('shared-ProductCard', module)
  .addDecorator(ContainerDeco)
  .add('default', () => <Story />)
  .add('notification', () => <Story notification={true} />)
  .add('subscribing', () => <Story subscribing={true} />)
  .add('summary', () => <Story summary={true} />);
