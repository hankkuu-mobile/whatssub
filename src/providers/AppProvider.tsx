import * as Font from 'expo-font';

import React, { useEffect, useReducer, useState } from 'react';

import { AsyncStorage } from 'react-native';
import { ThemeType } from '../types';

interface Context {
  state: State;
  changeTheme: (theme?: ThemeType) => void;
}
const AppContext = React.createContext<Context>(null);
const AppConsumer = AppContext.Consumer;

interface ChangeThemeTypeAction {
  type: 'change-theme-type-mode';
  payload: {
    theme?: ThemeType;
  };
}

interface Props {
  theme?: ThemeType;
  children?: any;
  doNotWaitFont?: boolean;
}

interface State {
  theme?: ThemeType;
}

export const initialState: State = {
  theme: ThemeType.LIGHT,
};

const reducer = (state: State, action: ChangeThemeTypeAction) => {
  switch (action.type) {
    case 'change-theme-type-mode':
      // if payload.theme doesn't exist, toggle
      if (!action.payload.theme) {
        return {
          ...state,
          theme:
            state.theme === ThemeType.LIGHT
              ? ThemeType.DARK
              : ThemeType.LIGHT,
        };
      }
      // else change it with payload.theme
      return {
        ...state,
        theme: action.payload.theme,
      };
    default:
      return state;
  }
};

function AppProvider(props: Props) {
  const [state, dispatch] = useReducer(reducer, initialState);
  const [fontLoading, setFontLoading] = useState(true);

  // by encapsulating dispatch, it is easier to use (such as passing through screenProps)
  const changeTheme: (theme?: ThemeType) => void = (theme) => {
    dispatch({
      type: 'change-theme-type-mode',
      payload: {
        theme,
      },
    });
  };

  // initialize theme
  useEffect(() => {
    changeTheme(props.theme ? props.theme : initialState.theme);
  }, [props.theme]);

  // Load font and then use saved theme from local storage
  const getCurrentThemeType = async () => {
    const value = (await AsyncStorage.getItem('theme')) as ThemeType;
    value && changeTheme(value);
  };
  const loadFonts = async () => {
    await Font.loadAsync({
      'spoqa-han-sans-bold': require('../../assets/fonts/SpoqaHanSans/SpoqaHanSans-Bold.ttf'),
      'spoqa-han-sans-regular': require('../../assets/fonts/SpoqaHanSans/SpoqaHanSans-Regular.ttf'),
    });

    setFontLoading(false);
  };
  useEffect(() => {
    loadFonts();
    getCurrentThemeType();
  }, []);

  useEffect(() => {
    AsyncStorage.setItem('theme', state.theme);
  }, [state.theme]);

  return (
    <AppContext.Provider
      value={{
        state,
        changeTheme,
      }}
    >
      {fontLoading && !props.doNotWaitFont ? null : props.children}
    </AppContext.Provider>
  );
}

export { AppConsumer, AppProvider, AppContext };
