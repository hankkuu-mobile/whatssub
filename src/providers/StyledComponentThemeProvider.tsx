import React, { useContext } from 'react';

import { AppContext } from './AppProvider';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import { createTheme } from '../theme';

const StyledComponentThemeProvider = (props) => {
  const { state } = useContext(AppContext);
  return (
    <ThemeProvider theme={createTheme(state.theme)}>
      {props.children}
    </ThemeProvider>
  );
};

StyledComponentThemeProvider.propTypes = {
  children: PropTypes.node,
};

export default StyledComponentThemeProvider;
