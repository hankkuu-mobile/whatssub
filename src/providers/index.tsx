import { AppConsumer, AppContext, AppProvider } from './AppProvider';

import PropTypes from 'prop-types';
import React from 'react';
import StyledComponentThemeProvider from './StyledComponentThemeProvider';
import { ThemeType } from '../types';

interface Props {
  theme?: ThemeType;
  children?: any;
  doNotWaitFont?: boolean;
}

const InitialProviders = ({ children, ...rest }: Props) => (
  <AppProvider {...rest}>
    <StyledComponentThemeProvider>{children}</StyledComponentThemeProvider>
  </AppProvider>
);

InitialProviders.propTypes = {
  children: PropTypes.node,
};

export { AppProvider, AppConsumer, AppContext, InitialProviders };
