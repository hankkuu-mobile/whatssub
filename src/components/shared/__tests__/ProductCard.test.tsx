import 'react-native';
import * as React from 'react';
import ProductCard, { ProductCardProps, Variant } from '../ProductCard';
import { RenderResult, act, fireEvent, render } from '@testing-library/react-native';

import { AppProvider } from '../../../providers';
import { ThemeProvider } from 'styled-components/native';
import { ThemeType } from '../../../types';
import { createTheme } from '../../../theme';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const props = {
  name: 'netflix',
  image: 'url',
  onClickEdit: () => {},
  isNotificationEnable: false,
  onClickNotification: () => {},
  price: 1234,
  currentMonthPaymentDate: new Date(2019, 8, 13),
};

const component = (props: ProductCardProps) => {
  return (
    <AppProvider doNotWaitFont>
      <ThemeProvider theme={createTheme(ThemeType.LIGHT)}>
        <ProductCard {...props}/>
      </ThemeProvider>
    </AppProvider>
  );
};

let rendered: renderer.ReactTestRendererJSON;
let testingLib: RenderResult;
// test for the container page in dom
describe('[ProductCard] ui rendering test', () => {
  it('should render outer component and snapshot matches', () => {
    rendered = renderer.create(component(props)).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('should include button when "variant" is subscription', () => {
    const Component = component({ ...props, variant: Variant.Subscription });
    rendered = renderer.create(Component).toJSON();
    testingLib = render(Component); // Todo: add SwitchToggle testID
  });

  it('should include summary title text when product card is pressed', () => {
    const Component = component(props);
    testingLib = render(Component);
    rendered = renderer.create(Component).toJSON();
    fireEvent.press(testingLib.getByTestId('productCard'));
    testingLib.getByTestId('summaryTitleText');
  });
});

describe('[ProductCard] Interaction', () => {
  let testingLib: RenderResult;

  it('should simulate [notification Icon] click', () => {
    let cnt = 0;
    const Component = component(
      { ...props, isNotificationEnable: true, onClickNotification: () => cnt++ });
    testingLib = render(Component);
    act(() => {
      fireEvent.press(testingLib.getByTestId('notiOffIcon'));
    });
    expect(cnt).toBe(1);
  });
});
