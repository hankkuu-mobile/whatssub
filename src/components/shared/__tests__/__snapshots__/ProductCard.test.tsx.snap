// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ProductCard] ui rendering test should render outer component and snapshot matches 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="productCard"
>
  <View
    isSummary={false}
    style={
      Array [
        Object {
          "backgroundColor": "white",
          "borderRadius": 15,
          "display": "flex",
          "height": 190,
          "paddingBottom": 15,
          "paddingLeft": 22,
          "paddingRight": 22,
          "paddingTop": 15,
          "position": "relative",
          "width": "100%",
        },
      ]
    }
  >
    <Image
      isSubscription={true}
      source={
        Object {
          "uri": "url",
        }
      }
      style={
        Array [
          Object {
            "height": 20,
            "marginBottom": 10,
            "marginLeft": "auto",
            "marginRight": "auto",
            "marginTop": 10,
            "width": 75,
          },
        ]
      }
      testID="productImg"
    />
    <View
      isSubscription={true}
      style={
        Array [
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "height": 40,
            "justifyContent": "space-between",
            "left": 22,
            "position": "absolute",
            "top": 15,
            "width": "100%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "flexDirection": "row",
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Image
            source={1}
            style={
              Array [
                Object {
                  "height": 24,
                  "width": 24,
                },
              ]
            }
          />
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="notiOnIcon"
        >
          <Image
            source={1}
            style={
              Array [
                Object {
                  "height": 24,
                  "width": 24,
                },
              ]
            }
          />
        </View>
      </View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(2, 48, 89, 1)",
              "borderRadius": 18,
              "flexDirection": "row",
              "height": 36,
              "padding": 3,
              "width": 72,
            }
          }
        >
          <View>
            <Text
              style={
                Object {
                  "color": "white",
                  "left": 5,
                  "position": "absolute",
                  "top": -7,
                }
              }
            >
              
            </Text>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "rgba(255, 255, 255, 1)",
                "borderRadius": 15,
                "height": 30,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
                "width": 30,
              }
            }
          >
            <View>
              <Text />
            </View>
          </View>
          <View>
            <Text
              style={
                Object {
                  "color": "white",
                  "left": 5,
                  "position": "absolute",
                  "top": -7,
                }
              }
            >
              Sub
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "display": "flex",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#323b43",
              "fontFamily": "spoqa-han-sans-bold",
              "fontSize": 16,
              "fontWeight": "bold",
              "marginTop": 17,
              "textAlign": "center",
            },
          ]
        }
      >
        netflix
      </Text>
      <Text
        isSummary={false}
        style={
          Array [
            Object {
              "color": "rgb(255, 77, 108)",
              "fontFamily": "spoqa-han-sans-regular",
              "fontSize": 40,
              "textAlign": "center",
            },
          ]
        }
        variant="subscription"
      >
        D-9
      </Text>
      <Text
        style={
          Array [
            Object {
              "fontFamily": "spoqa-han-sans-bold",
              "fontSize": 13,
              "textAlign": "center",
            },
          ]
        }
      >
        9월 13일에
                  1,234원이 결제될 예정입니다.
      </Text>
    </View>
  </View>
</View>
`;
